### PR TITLE
Consume new DevSiteRoutingMenu in generateRouteConfig

### PR DESCRIPTION
**Pull request by cody-dot-js**
### Summary

This adds the ability for the navigation side menu to filter/search for components, using the toolbar from [this terra-framework PR#352](https://github.com/cerner/terra-framework/pull/352)

#### Example Usage GIF:

@cerner/terra

[CONTRIBUTORS.md]: https://github.com/cerner/terra-dev-site/blob/master/CONTRIBUTORS.md
[License]: https://github.com/cerner/terra-dev-site/blob/master/License.md

### CODE REVIEW

**Comment by dev-cprice** [comment]
This will need to wait until https://github.com/cerner/terra-framework/pull/352 is merged and released. Opening for review until then.

**Comment by dev-cprice** [comment]
Ready for review now.

File: src/app/components/DevSiteRoutingMenu.jsx [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+const propTypes = {
+  /**
+   * The title to render within the DevSiteRoutingMenu's header.
+   */
+  title: PropTypes.string,
+  /**
+   * The array of routing shapes to be rendered as menu items within the DevSiteRoutingMenu.
+   */
+  menuItems: PropTypes.arrayOf(PropTypes.shape({
```
Comment (yuderekyu) [inline src/app/components/DevSiteRoutingMenu.jsx]: i know this was a copy + paste, but this propType is defined in `application-layout`. we can import that:
``` jsx
import { Utils } from 'terra-application-layout';
...
...
menuItems: Utils.propTypes.layoutConfigPropType
```

File: src/app/components/DevSiteRoutingMenu.jsx [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    text: PropTypes.string.isRequired,
+    path: PropTypes.string.isRequired,
+    hasSubMenu: PropTypes.bool,
+  })),
+  /**
+   * The Object of layout-related APIs provided to the components of the Layout.
+   */
+  layoutConfig: PropTypes.shape({
```
Comment (yuderekyu) [inline src/app/components/DevSiteRoutingMenu.jsx]: same here

File: src/app/components/DevSiteRoutingMenu.jsx [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+      childKeys: processedMenuItems.map(item => item.key),
+      isRootMenu: !routingStackDelegate.showParent && !title,
+    });
+
+    const hasToolbar = (items || []).length >= FILTER_DISPLAY_THRESHOLD;
+    const toolbar = (hasToolbar
+      ? (
+        <FormattedMessage id="Terra.devSiteRoutingMenu.filter">
+          {placeholder => (
+            <SearchField
```
Comment (emilyrohrbough) [inline src/app/components/DevSiteRoutingMenu.jsx]: Do we want to toggle rendering when a SearchField is available or do we want to disable the SearchField when the number of menu items are less than the threshold?

File: src/app/components/DevSiteRoutingMenu.jsx [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+  },
+}));
+
+/**
+ * This function compares the given path against the paths of the given menuItems. If a match
+ * (partial or otherwise) is detected, that path is returned. If no match is detected, `undefined` is returned.
+ * @param {String} pathToMatch is the path to be matched against
+ * @param {Array} menuItems is the Array of menuItem objects as specified by the DevSiteRoutingMenu's proptype definition.
+ */
+const getSelectedChildKey = (pathToMatch = '', menuItems = []) => (menuItems.filter(({ path }) => path && matchPath(pathToMatch, { path }))[0] || {}).path;
```
Comment (emilyrohrbough) [inline src/app/components/DevSiteRoutingMenu.jsx]: This component is almost one-to-one with the terra-application-layout's RoutingMenu. Should we match some of the implementation? Like maintaining the `getSelectableChildren` & `buildSideMenuItems` as static?
Reference: [getSelectableChildren](https://github.com/cerner/terra-framework/blob/7fb36175aa4c9ec52f1d80255041b08777ffa46c/packages/terra-application-layout/src/menu/RoutingMenu.jsx#L90) && [buildSideMenuItems](https://github.com/cerner/terra-framework/blob/7fb36175aa4c9ec52f1d80255041b08777ffa46c/packages/terra-application-layout/src/menu/RoutingMenu.jsx#L48)

File: src/app/components/DevSiteRoutingMenu.jsx [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    /**
+     * The selectedChildKey is re-evaluated when new props are received to keep the internal state consistent.
+     */
+    this.setState({
+      selectedChildKey: getSelectedChildKey(nextProps.location.pathname, nextProps.menuItems),
+      items: nextProps.menuItems,
```
Comment (emilyrohrbough) [inline src/app/components/DevSiteRoutingMenu.jsx]: nit: items -> menuItems

**Comment by emilyrohrbough** [comment]
Looks like we are missing fi-FI translations

File: src/app/components/DevSiteRoutingMenu.jsx (hunk 1 above)
Comment (dev-cprice) [reply src/app/components/DevSiteRoutingMenu.jsx]: 👍 I originally had this, but the linter complained. I'll add it back with an eslint-disable

File: src/app/components/DevSiteRoutingMenu.jsx (hunk 2 above)
Comment (dev-cprice) [reply src/app/components/DevSiteRoutingMenu.jsx]: Same as above

File: src/app/components/DevSiteRoutingMenu.jsx (hunk 3 above)
Comment (dev-cprice) [reply src/app/components/DevSiteRoutingMenu.jsx]: I'm not sure I understand the question..

UX wants the filter to only appear when there are at least 10 items in the nav

File: src/app/components/DevSiteRoutingMenu.jsx (hunk 4 above)
Comment (dev-cprice) [reply src/app/components/DevSiteRoutingMenu.jsx]: What is the benefit? That makes the implementation public on the object.. do we want that?

File: src/app/components/DevSiteRoutingMenu.jsx (hunk 3 above)
Comment (emilyrohrbough) [reply src/app/components/DevSiteRoutingMenu.jsx]: As you navigate through nested links, does this change when the search field is displayed?

Comment (emilyrohrbough) [reply src/app/components/DevSiteRoutingMenu.jsx]: For example,
Terra-Core ~= 80 list items, show search field 
> click terra-list
List ~= 3, don't show search field

File: src/app/components/DevSiteRoutingMenu.jsx (hunk 4 above)
Comment (emilyrohrbough) [reply src/app/components/DevSiteRoutingMenu.jsx]: Benefit that implementation would be consistent incase an enhancement was made or a bug was found in the terra-applicaiton-layout's implementation and we need to make the same changes here. However, I'm unsure about it making the implementation public. @tbiethman @dkasper-was-taken do you know?

File: src/app/components/DevSiteRoutingMenu.jsx (hunk 3 above)
Comment (dev-cprice) [reply src/app/components/DevSiteRoutingMenu.jsx]: Correct, we don't want to show the search field in the terra-list..

I initially had a typo, but the implementation is to only display the filter iff the initial menu items (`{ menuItems } = this.props`) length >= 10

File: src/app/components/DevSiteRoutingMenu.jsx (hunk 4 above)
Comment (bjankord) [reply src/app/components/DevSiteRoutingMenu.jsx]: I'm okay with this as-is.

**Review by bjankord — APPROVED** [review]

**Comment by emilyrohrbough** [comment]
Could you open a PR in core so we could see what this would look like?

**Comment by dev-cprice** [comment]
@emilyrohrbough I can, but it won't be noticeable until this PR and https://github.com/cerner/terra-framework/pull/352 are merged and released. I'll open a PR to core with the config change needed and then `npm-link` all the necessary parts and record a gif though.

**Comment by emilyrohrbough** [comment]
Okay cool! I am curious to see what happens while you navigate through with changing list lengths

**Comment by dev-cprice** [comment]
@emilyrohrbough PR with demo: https://github.com/cerner/terra-core/pull/2108 🙌

**Pull request merged**